core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -186,8 +185,8 @@ object Uri extends UriFunctions {
 
 
   private def renderParamsAndFragment(writer: Writer, p: Query, f: Option[Fragment]): writer.type = {
-    if (p.nonEmpty) writer << '?' << p
-    if (f.isDefined) writer << '#' << UrlCodingUtils.urlEncode(f.get, spaceIsPlus = false)
+    if (!p.isEmpty) writer << p
```

Comment:
It moved into the query itself so that the `EmptyQuery` (which is distinct from the non-empty zero-length query) can omit it.  I can't remember whether `Option[Query]` would work alright.